### PR TITLE
Version 1.0

### PR DESCRIPTION
**Pull request by cliffckerr**
### Description
- Official release of Starsim!
- Adds a ``Calibration`` class, based on Optuna, to facilitate the calibration of Starsim models.
- Adds ``mean()``, ``median()``, and ``plot()`` methods to ``MultiSim``.
- Adds ``low`` and ``high`` attributes to ``Result`` objects.
- Adds a ``flatten()`` method to ``Results``, allowing nested ``Results`` objects to be turned into flat dictionaries.
- Removes duplicate UIDs among new infections, and adds a ``unique()`` method to ``ss.uids``.
- Fixes a bug that prevented ``ss.lognorm_im()`` from using callable parameters.
- Updates the default ``Sim`` string representation to be a single line; the more verbose version is available via ``sim.disp()``.

### Checklist
- [x] Code commented & docstrings added
- [x] New tests were needed and have been added
- [x] A new version number was needed & changelog has been updated
- [x] A new PyPI version needs to be released

### CODE REVIEW

File: README.rst [hunk 1]
```diff
@@ -1,13 +1,13 @@
...
+Examples of diseases that have already been implemented in Starsim include sexually transmitted infections (HIV, HPV, syphilis, gonorrhea, and chlamydia, including co-transmission), respiratory infections (tuberculosis and RSV), plus other diseases (Ebola and cholera) and underlying determinants of health (such as malnutrition).
```
Comment (robynstuart) [inline README.rst:6]: i would say gonorrhea and chlamydia aren't really done, and can't be done well until the time refactor is in place

File: starsim/disease.py [hunk 2]
```diff
@@ -295,8 +295,9 @@ def make_new_cases(self):
         # Tidy up
         if len(new_cases) and len(sources):
             new_cases = ss.uids.cat(new_cases)
-            sources = ss.uids.cat(sources)
-            networks = np.concatenate(networks)
+            new_cases, inds = new_cases.unique(return_index=True)
+            sources = ss.uids.cat(sources)[inds]
```
Comment (robynstuart) [inline starsim/disease.py:299]: does this guarantee that len(new_cases) = len(sources)?

**Review by robynstuart — COMMENTED** [review]
Looks great!!!

**Comment by robynstuart** [comment]
Playing around with the new repr:
```
sir = ss.SIR(init_prev=ss.bernoulli(0.3))
sim = ss.Sim(diseases=sir)
/Users/robynstuart/Documents/git/starsim/starsim/sim.py:72: RuntimeWarning: 
Error displaying custom sim repr, falling back to default: object of type 'SIR' has no len()
  ss.warn(f'Error displaying custom sim repr, falling back to default: {E}')
```

File: starsim/calibration.py [hunk 3]
```diff
@@ -0,0 +1,496 @@
...
+        df_res = sim.export_df()
+        df_res['year'] = np.floor(np.round(df_res.index, 1)).astype(int)
+        for skey in self.sim_result_list:
+            if 'prevalence' in skey:
+                model_output = df_res.groupby(by='year')[skey].mean()
+            else:
+                model_output = df_res.groupby(by='year')[skey].sum()
+
+            data = self.target_data[skey]
+            combined = pd.merge(data, model_output, how='left', on='year')
```
Comment (daniel-klein) [inline starsim/calibration.py:310]: Could provide `suffixes` instead of accepting the default of `_x` and `_y` to differentiate data from model.

File: starsim/calibration.py [hunk 4]
```diff
@@ -0,0 +1,496 @@
...
+            combined['diffs'] = combined[skey+'_x'] - combined[skey+'_y']
+            gofs = compute_gof(combined.dropna()[skey+'_x'], combined.dropna()[skey+'_y'])
+
+            losses = gofs  #* self.weights[skey]
+            mismatch = losses.sum()
```
Comment (daniel-klein) [inline starsim/calibration.py:315]: AH, so if gof returns a timeseries... the values get summed here. So maybe change the default `none` to `sum` so that it's more self contained in the gof function?

File: starsim/calibration.py [hunk 5]
```diff
@@ -0,0 +1,496 @@
...
+        op = import_optuna()
+        if not self.run_args.keep_db:
+            self.remove_db()
+        if self.run_args.rand_seed is not None:
+            sampler = op.samplers.RandomSampler(self.run_args.rand_seed)
+            sampler.reseed_rng()
+            raise NotImplementedError('Implemented but does not work')
+        else:
+            sampler = None
+        print(self.run_args.storage)
```
Comment (daniel-klein) [inline starsim/calibration.py:372]: Is this `print` intentional?

Comment (daniel-klein) [reply starsim/calibration.py:372]: (If so, it could be more verbose)

File: tests/test_calibration.py [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+        beta = {'random': [0.01]*2, 'maternal': [1, 0]},
+        init_prev = 0.15,
+    )
+    pregnancy = ss.Pregnancy(fertility_rate=20)
+    death = ss.Deaths(death_rate=10)
+    random = ss.RandomNet(n_contacts=4)
+    maternal = ss.MaternalNet()
+
+    sim = ss.Sim(
+        dt = 1,
```
Comment (daniel-klein) [inline tests/test_calibration.py:27]: Try some other `dt` to improve the test.

File: tests/test_calibration.py [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+    [  2015,  13814642,        0.0941030,          1300000,               44000 ,           24000],
+    [  2016,  14030338,        0.0926563,          1300000,               43000 ,           23000],
+    [  2017,  14236599,        0.0913139,          1300000,               34000 ,           23000],
+    [  2018,  14438812,        0.0900351,          1300000,               27000 ,           22000],
+    [  2019,  14645473,        0.0920401,          1347971,               23000 ,           None],
+    [  2020,  14862927,        0.0874659,          1300000,               20000 ,           None],
+    [  2021,  15085870,        0.0861733,          1300000,               19000 ,           None],
+    [  2022,  15312158,        0.0848998,          1300000,               17000 ,           None],
+    ]
+    df = sc.dataframe(target_data[1:], columns=target_data[0])
```
Comment (daniel-klein) [inline tests/test_calibration.py:78]: Would prefer a straghtforward pd.DataFrame here unless really leveraging the extension offered in sciris.

Comment (daniel-klein) [reply tests/test_calibration.py:78]: Could be:
```python
df = pd.DataFrame({
    'n_alive': [10432409, 10681008, ...],
    'hiv.prevalence': [0.0699742, 0.0851979, ...],
    ...
}, index=pd.Index([1990, 1991, ...], name='year'))
```

File: starsim/calibration.py [hunk 8]
```diff
@@ -0,0 +1,496 @@
...
+        return sim
+
+    def trial_to_sim_pars(self, pardict=None, trial=None):
+        """
+        Take in an optuna trial and sample from pars, after extracting them from the structure they're provided in
+        Different use cases:
+            - pardict is self.calib_pars, i.e. {'diseases':{'hiv':{'art_efficacy':[0.96, 0.9, 0.99]}}}, need to sample
+            - pardict is self.initial_pars, i.e. {'diseases':{'hiv':{'art_efficacy':[0.96, 0.9, 0.99]}}}, pull 1st vals
+            - pardict is self.best_pars, i.e. {'diseases':{'hiv':{'art_efficacy':0.96786}}}, pull single vals
+        """
```
Comment (daniel-klein) [inline starsim/calibration.py:231]: Could we add a callable mapping? Lots of reasons to do this as calibration parameters _often_ do not align 1:1 to model inputs. Edge case worth consideration is distributions with callable parameters.

File: tests/test_calibration.py [hunk 9]
```diff
@@ -0,0 +1,148 @@
...
+    # Define the calibration parameters
+    calib_pars = dict(
+        diseases = dict(
+            hiv = dict(
+                init_prev = [0.15, 0.01, 0.30],
+            ),
+        ),
+        networks = dict(
+            randomnet = dict(
+                n_contacts = [4, 2, 10],
```
Comment (daniel-klein) [inline tests/test_calibration.py:95]: What if these were `ss.poisson` instead of scalars?

File: starsim/results.py [hunk 10]
```diff
@@ -18,14 +18,18 @@ class Result(np.ndarray):
         'module', # The name of the module (e.g. hiv)
         'scale', # Whether or not the result scales with population size (e.g. True)
         'label', # The human-readable label for the result (e.g. Number of infections)
+        'low', # The lower bound for the values (used with MultiSim)
```
Comment (daniel-klein) [inline starsim/results.py:21]: Mention these low and high values are used for plotting confidence intervals. If not used for any other purpose, perhaps better to move into the `MultiSim` class?

File: README.rst [hunk 11]
```diff
@@ -1,13 +1,13 @@
...
 
-The framework is appropriate for simulating sexually transmitted infections (including syphilis, gonorrhea, chlamydia, HPV, and HIV, including co-transmission), respiratory infections (like RSV and tuberculosis), and other diseases and underlying determinants (such as Ebola, diabetes, and malnutrition).
+Note: Starsim is part of the same suite of tools as `Covasim <https://covasim.org>`_, `HPVsim <https://hpvsim.org>`_, and `FPsim <https://fpsim.org>`_, which are collectively known as the "Starism suite".
```
Comment (daniel-klein) [inline README.rst:8]: Perhaps be more clear that Starsim (this repo) is a general purpose framework building on previous disease-specific models...

File: README.rst [hunk 12]
```diff
@@ -29,32 +29,38 @@ Starsim can also be installed locally. To do this, clone first this repository,
 Usage and documentation
 -----------------------
 
-Documentation, including tutorials and an API reference, is available at https://docs.starsim.org. 
+Full documentation, including tutorials and an API reference, is available at https://docs.starsim.org. 
 
-If everything is working, the following Python commands will run a simulation with the simplest version of a Starsim model. We'll make a version of a classic SIR model::
+If everything is working, you can run a simple demo via::
```
Comment (daniel-klein) [inline README.rst:34]: Remove "If everything is working"

File: README.rst [hunk 13]
```diff
@@ -29,32 +29,38 @@ Starsim can also be installed locally. To do this, clone first this repository,
...
-    ),
-    diseases = dict(    # *Diseases* add detail on what diseases to model
-      type = 'sir',     # Here, we're creating an SIR disease
-      init_prev = 0.1,  # Proportion of the population initially infected
-      beta = 0.5,       # Probability of transmission between contacts
-    )
+      n_agents = 5_000,     # Number of agents to simulate
+      networks = dict(      # Networks define how agents interact w/ each other
+          type = 'random',  # Here, we use a 'random' network
+          n_contacts = 10   # Each person has an average of 10 contacts w/ other people
```
Comment (daniel-klein) [inline README.rst:48]: If scalar, all agents should receive exactly 10 contacts (although I do believe self contacts are possible with the current implementation).

**Review by daniel-klein — COMMENTED** [review]
Exiting to be so close to 1.0! I've added lots of comments in the new calibration code. Other changes seem great, just some minor suggestions here and there. Calibration changes will likely get push post 1.0 release.

File: starsim/disease.py (hunk 2 above)
Comment (cliffckerr) [reply starsim/disease.py:299]: it should!

File: README.rst (hunk 1 above)
Comment (cliffckerr) [reply README.rst]: fixed

File: starsim/distributions.py [hunk 14]
```diff
@@ -661,8 +661,10 @@ def __init__(self, mean=0.0, sigma=1.0, **kwargs):
         super().__init__(distname='lognormal', dist=sps.lognorm, mean=mean, sigma=sigma, **kwargs)
         return
     
-    def sync_pars(self):
+    def sync_pars(self, call=True):
         """ Translate between NumPy and SciPy parameters """
+        if call:
+            self.call_pars() # Since can't work with functions
```
[the remark this answers is not in the document]
Comment (cliffckerr) [reply starsim/distributions.py:667]: fixed (outdated comment)

File: starsim/network.py [hunk 15]
```diff
@@ -902,15 +902,14 @@ def add_pairs(self):
         loc = self.pars.embedding_func.rvs(available)
         loc_f = loc[people.female[available]]
         loc_m = loc[~people.female[available]]
-
+
```
[the remark this answers is not in the document]
Comment (cliffckerr) [reply starsim/network.py:905]: fixed

File: README.rst (hunk 11 above)
Comment (cliffckerr) [reply README.rst]: fixed
